imgtops.1
.TH IMGTOPS "1" "September 2003" "imgtops 1.0" "User Commands"
.SH NAME
imgtops \- convert images to PostScript
.SH SYNOPSIS
imgtops [options]... [filename(s)]...
.SH DESCRIPTION
This program reads one or more images in any of a wide variety of formats
(including JPEG, PNG, GIF, Targa, TIFF, BMP, and many others) and
produces a Encapsulated PostScript Level 2 or 3 file as
output.  The goal is to produce the most space-efficient encoding of
the image.
.PP
The files output will print only on PostScript devices
supporting Level 2 features, but Level 2 has been around since 1991 so
virtually all printers today support it.
.PP
If no input file is specified, \fBimgtops\fR reads from stdin.
.PP
The output of this program with no options will be a one-page
PostScript document with the image centered on a US letter-size page
with one-inch margins.  The file
will contain a tight bounding box for the image and other necessary
DSC comments to make the output a legal Encapsulated PostScript (EPS)
file, suitable for inclusion in other documents.
.SH OPTIONS
.TP
\fB\-?\fR, \fB\-\-help\fR
Causes the program to show an option summary and usage information, then exit.
.TP
\fB\-\-version\fR
Causes the program to show version and license information, then exit.
.TP
\fB\-e\fR, \fB\-\-strict\-eps\fR
The normal output of \fBimgtops\fR is a page with a centered image.  The
exact size and position depends on the various page size, image size,
and margin parameters listed below.  The document always contains a tight
EPS bounding box around the image, but since bounding boxes must have
integral coordinates, there may be small slivers of empty page
contained in the bounding box as well.  The \fB\-e\fR option makes the
bounding box fit the image \fIexactly,\fR at the expense of centering
it on the page.  These files are suitable only for EPS inclusion in
other documents; parts of the image will likely be cut off if they are
printed directly.  \fB\-e\fR causes the image size, page size, and page
margin options to be ignered.   

The summary is:  use \fB\-e\fR if you're only using the output file by
including it in another document.  Don't use \fB\-e\fR if you want to
print the output file by itself.
.TP
\fB\-w\fR, \fB\-\-width=\fR<dim>
.TP
\fB\-h\fR, \fB\-\-height=\fR<dim>
Set the desired image width and height.  See the DIMENSIONS
section below for legal dimension values.  If both of these options
are given, the image may be scaled nonuniformly.  If the result does
not fit on the page minus the margins, the image is scaled
down to fit while maintaining its aspect ratio.  The default if
neither is given is to make the image as large as possible.
.TP
\fB\-s\fR, \fB\-\-paper\-size\fR=\fISIZE\fR
Set the output page size.  The size value may be "letter", "legal", or
"a4" to select one of the builtin sizes, or it may be two dimensions
(width and height) separated by a comma.
.TP
\fB\-m\fR, \fB\-\-margin=\fR<dim>
Sets the size of the page margin.  The default is one inch.
.TP
\fB\-l\fR, \fB\-\-landscape\fR
Rotates the image 90 degrees on the page.  This may be used in
conjunction with \fB\-e\fR.
.TP
\fB\-2\fR, \fB\-\-level\-2\fR
Forces all output to be Level 2 PostScript.  This is the default.  If
both \fB\-3\fR and \fB\-2\fR are given, the one that appears last
takes effect.
.TP
\fB\-3\fR, \fB\-\-level\-3\fR
Allows the use of Level 3 PostScript features, which may result in
smaller output files.
.TP
\fB\-8\fR, \fB\-\-allow\-8\-bit\fR
By default the output file is 7-bit clean, containing only printable
ASCII characters and whitespace.  This option allows arbitrary binary
data to appear in the output, which reduces the size of the output
file by about 20%.  However, some PostScript document managers (such as \fIgv\fR, as of this
writing) choke on binary data.  For maximum portability of the
resulting PostScript, do not use this option.
.TP
\fB\-x\fR, \fB\-\-hex\-encoding\fR
Uses a hexadecimal encoding of binary data instead of the default
ASCII85 encoding.  This option is primarily for debugging -- it
results in an output file that is 60% larger, with no corresponding
benefit.  This option has no effect if \fB\-8\fR is specified (since
then the binary data is not encoded at all, but appears in the output file directly).
.TP
\fB\-n\fR, \fB\-\-line\-length=\fR#
By default, ASCII-encoded data is broken into lines of 75 characters
or less.  The \fB\-n\fR option changes that limit.  Setting it to zero
or a negative value disables line breaking; the output will be one
very long line.  This option effects only the image data portion of
the output file.  The header of a typical output file contains lines
of around 75 characters; \fB\-n\fR has no effect on these lines.  This
option has no effect when \fB\-8\fR is specified.
.TP
\fB\-b\fR, \fB\-\-batch\fR
In batch mode, \fBimgtops\fR accepts multiple input images and
processes them all.  It outputs one file per image; the output
filenames are constructed by replacing the input file's extension with
".eps" (if \fB\-e\fR is used) or ".ps" (if it is not).  The \fB\-o\fR
option can be used to specify the directory where the output files
should be placed.

At least one input image filename must be given; batch input images can not
come from stdin.
.TP
\fB\-o\fR, \fB\-\-output\fR <filename>
.TP
\fB\-o\fR, \fB\-\-output\fR <dirname>
Normally \fBimgtops\fR processes a single input image, and \fB\-o\fR
is used to specify the output filename.  The default is stdout.

In batch mode (selected with \fB\-b\fR), this option specifies a
\fIdirectory\fR for the output files to go into; the filenames are
constructed automatically from the input filenames.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Normally \fBimgtops\fR prints a message to stderr as each input file
is processed.  \fB\-q\fR suppresses all output to stderr, except for
actual error messages.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
The opposite of \fB\-q\fR, this causes additional information to be
printed to stderr as the program runs.
.SH DIMENSIONS
Any option requiring a dimension may be given as a number followed by
"in", "cm", "mm", "pt", "ft", or "m" to indicate the units.  A number
given without units is assumed to be a distance in PostScript points
(1/72-inch).
.SH COPYRIGHT
Copyright \(co 2003 Doug Zongker
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.
.PP
PostScript is a registered trademark of Adobe Systems Incorporated.



CHANGES
version 1.0    :  4 Sep 2003

   PIL's handling of paletted images changed in version 1.1.4; imgtops
   now handles both the new and older versions.  Haven't run across
   any major bugs since March, so we'll call it 1.0!
	
version 0.9.2  :  24 Mar 2003

   imgtops handles most PIL image modes via conversion.  (Notably,
   RGBA images are handled by dropping the alpha channel.)

   ASCII data line lengths are limited to 75 characters by default.
   Added -n option to change this.

   Level 3 output files will now print on level 2 devices, but they
   will show a grey box and a numeral '3' in place of the image.

version 0.9.1  :   2 Mar 2003

   Added rotation option to epstoimg.  Changed short form of
   "resolution" option to "-s".

version 0.9    :  22 Feb 2003

   First released version.


imgtopslib/imgtops.py
import sys, os.path
import Image
import cStringIO
import time
import getopt
from dimensions import interpret_dimension
import psimage

version = 'imgtops 1.0'
copyright = '''Copyright (C) 2003 Doug Zongker
This is free software; see the accompanying license for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.'''



timestamp = time.asctime( time.localtime() )

paper_sizes = {
    'letter' : (612, 792),
    'legal' : (612, 1008),
    'a4' : (595, 849),
    }

err = sys.stderr

class Parameters:
    def __init__( self ):
        self.level = 2
        self.force_hex = 0
        self.skip_rle = 0
        self.verbose = 10
        self.strict_eps = 0
        self.binary_output = 0
        self.line_length = 75
        
        self.width = None
        self.height = None
        self.paper_size = paper_sizes['letter']
        self.margin = interpret_dimension( '1in' )
        self.landscape = 0
        
        self.batch = 0
        self.input = None
        self.output = None

    def dump( self, f ):
        print >> f, 'level %d  force_hex %d  skip_rle %d  strict_eps %d  verbose %d' % \
              (self.level, self.force_hex, self.skip_rle, self.strict_eps, self.verbose)
        print >> f, 'width %r  height %r' % (self.width, self.height)
        print >> f, 'paper_size %r  margin %.2f  landscape %d' % (self.paper_size, self.margin, self.landscape)
        print >> f, 'batch %d' % (self.batch,)
        print >> f, 'inputs:', self.input
        print >> f, 'outputs:', self.output
        

params = Parameters()

def usage():
    print 'usage: %s [options] [filename(s)]' % (sys.argv[0],)
    print 
    print '  -?, --help                   show this message and exit'
    print '  --version                    print version info and exit'
    print 
    print '  -w, --width=<dim>            set image width'
    print '  -h, --height=<dim>           set image height'
    print '  -s, --paper-size=SIZE        set page size'
    print '        available sizes:         letter, legal, a4, or "<dim>,<dim>"'
    print '        default:                 letter'
    print '  -l, --landscape              use landscape mode'
    print '  -m, --margin=<dim>           set page margin (default 1 inch)'
    print
    print '        example dimensions:      4in, 35mm, 30.1cm, 144pt'
    print
    print '  -e, --strict-eps             strict EPS bounding box'
    print '  -2, --level-2                only level 2 postscript (default)'
    print '  -3, --level-3                allow level 3 postscript'
    print '  -x, --hex-encoding           force hex encoding'
    print '  -8, --allow-8-bit            allow binary data in output'
    print '  -n, --line-length=#          limit ASCII data lines to # characters'
    print 
    print '  -b, --batch                  accept multiple input images'
    print '  -o, --output <filename>      output file name (non-batch mode)'
    print '  -o, --output <dirname>       output directory (batch mode)'
    print
    print '  -q, --quiet                  run silently, except for errors'
    print '  -v, --verbose                be verbose'
    

def showversion():
    print version
    print
    print copyright
    

def main():
    try:
        opts, args = getopt.getopt(sys.argv[1:], '?w:h:ls:m:e238xbo:qvn:',
                                   ['width=',
                                    'height=',
                                    'paper-size=',
                                    'landscape',
                                    'margin=',
                                    'strict-eps',
                                    'level-2',
                                    'level-3',
                                    'allow-8-bit',
                                    'line-length=',
                                    'hex-encoding',
                                    'batch',
                                    'output=',
                                    'quiet',
                                    'verbose',
                                    'version',
                                    ])
    except getopt.GetoptError:
        usage()
        return 2
    output = None
    for o, a in opts:
        if o in ('-?', '--help'):
            usage()
            return 0
        elif o == '--version':
            showversion()
            return 0
        elif o in ('-v', '--verbose'):
            params.verbose = 20
        elif o in ('-q', '--quiet'):
            params.verbose = 0
        elif o in ('-w', '--width'):
            params.width = interpret_dimension(a)
            if params.width is None:
                print >> err, "didn't understand width dimension %r" % (a,)
                return 1
            elif params.width <= 0:
                print >> err, "image width must be positive"
                return 1
        elif o in ('-h', '--height'):
            params.height = interpret_dimension(a)
            if params.height is None:
                print >> err, "didn't understand height dimension %r" % (a,)
                return 1
            elif params.height <= 0:
                print >> err, "image height must be positive"
                return 1
        elif o in ('-s', '--paper-size'):
            params.paper_size = interpret_papersize(a)
            if params.paper_size is None:
                print >> err, "didn't understand paper size %r" % (a,)
                return 1
        elif o in ('-m', '--margin'):
            params.margin = interpret_dimension(a)
            if params.margin is None:
                print >> err, "didn't understand margin dimension %r" % (a,)
                return 1
            elif params.margin < 0:
                print >> err, "margin must be nonnegative"
                return 1
        elif o in ('-l', '--landscape'):
            params.landscape = 1
        elif o in ('-e', '--strict-eps'):
            params.strict_eps = 1
        elif o in ('-x', '--hex-encoding'):
            params.force_hex = 1
        elif o in ('-2', '--level-2'):
            params.level = 2
        elif o in ('-3', '--level-3'):
            params.level = 3
        elif o in ('-8', '--allow-8-bit'):
            params.binary_output = 1
        elif o in ('-n', '--line-length'):
            try:
                params.line_length = int(a)
            except ValueError:
                print >> err, "line length must be integer"
                return 1
        elif o in ('-b', '--batch'):
            params.batch = 1
        elif o in ("-o", "--output"):
            params.output = a

    if not params.strict_eps and \
           (params.margin * 2 >= params.paper_size[0] or
            params.margin * 2 >= params.paper_size[1]):
        print >> err, "margin leaves no room on page for image"
        return 1

    if params.batch and len(args) == 0:
        print >> err, "image filenames must be specified for batch mode"
        return 1

    if not params.batch and len(args) > 1:
        print >> err, "multiple input filenames given; perhaps use batch mode?"
        return 1

    if params.batch:
        params.input = args
        if params.output is None:
            params.output = ''
        else:
            if not os.path.isdir( params.output ):
                print >> err, "in batch mode, -o must specify a directory"
                return 1
    else:
        if args:
            params.input = args[0]

    if params.batch:
        run_batch_mode()
    else:
        run_single_mode()


def interpret_papersize( a ):
    try:
        return paper_sizes[a]
    except KeyError:
        pass
    
    wh = a.split(',')
    if len(wh) != 2:
        return None

    try:
        w = interpret_dimension(wh[0])
        h = interpret_dimension(wh[1])
    except ValueError:
        return None

    if w > 0 and h > 0:
        return w, h
    else:
        return None


def run_single_mode():
    if params.verbose >= 10:
        if params.input is None:
            print >> err, "processing [<stdin>]..."
        else:
            print >> err, "processing [%s]..." % (params.input,)
    
    try:
        lim = load_image( params.input )
    except IOError, msg:
        print >> err, "  image read error: %s" % (msg,)
        return 1

    if params.output:
        try:
            f = open( params.output, 'wb' )
        except IOError, msg:
            print >> err, "  PS write error: %s" % (msg,)
            return 1
    else:
        f = sys.stdout

        # change stdout's mode to binary, to avoid
        # newlines getting translated to CRLFs.
        
        if sys.platform == "win32":
            import os, msvcrt
            msvcrt.setmode(sys.stdout.fileno(), os.O_BINARY)

    process_one_image( lim, f )
    f.close()
    return 0
            
def run_batch_mode():
    if params.strict_eps:
        newext = '.eps'
    else:
        newext = '.ps'
    result = 0
        
    for i in params.input:
        if params.verbose >= 10:
            print >> err, "processing [%s]..." % (i,)
            
        try:
            lim = load_image( i )
        except IOError, msg:
            print >> err, "  image read error: %s; skipping" % (msg,)
            result = result or 1
            del lim
            continue

        dir, name = os.path.split( i )
        base, ext = os.path.splitext( name )
        outpath = os.path.join( params.output, base+newext )

        try:
            f = open( outpath, 'wb' )
        except IOError, msg:
            print >> err, "  PS write error: %s; skipping" % (msg,)
            result = result or 1
            continue

        process_one_image( lim, f )
        del lim
        f.close()

    return result

def process_one_image( lim, f ):
    d = { 'Creator' : '(%s)' % (version),
          'CreationDate' : '(%s)' % (timestamp,),
          'Title' : '(%s)' % (lim.filename,) }

    imbuffer = cStringIO.StringIO()
    level = psimage.write_ps_image( imbuffer, lim, params )
    
    fit = psimage.compute_fit( lim.size, params )
    psimage.write_postscript_header( fit, params, f, d, level )
    f.write( imbuffer.getvalue() )
    imbuffer.close()
    psimage.write_postscript_footer( fit, params, f )
    


class LoadedImage:
    pass
            
def load_image( filename ):
    if filename == None:
        buffer = cStringIO.StringIO( sys.stdin.read() )
    else:
        f = open( filename, 'rb' )
        buffer = cStringIO.StringIO( f.read() )
        f.close()
    im = Image.open( buffer )

    lim = LoadedImage()
    lim.buffer = buffer
    lim.im = im
    lim.size = im.size
    if filename:
        lim.filename = filename
    else:
        lim.filename = '<stdin>'

    return lim

if __name__ == "__main__":
    r = main()
    sys.exit( r )


# lim = load_image( sys.argv[1] )
# f = open( 'test.eps', 'w' )
# write_ps_image( f, lim )
# f.close()
